!!! BUGFIX: DocumentUriProjection doesn't respect fallback nodes; NodePropertiesWereSet contains affectedDimensionSpacePoints

skurfuerst opened the pull request:
Extending the NodePropertiesWereSet event is useful for some projections which are not interested in OriginDimensionSpacePoints, but need to know where a property value is applied (like the DocumentUriPathProjection).

This change is BREAKING because it updates the event payload of NodePropertiesWereSet. To update your event store, run:

./flow migrateEvents:fillAffectedDimensionSpacePointsInNodePropertiesWereSet

We also fix the DocumentUriPathProjection alongside.

Resolves: #4265
Resolves: #4256
Resolves: #4270 
Resolves: #4253

skurfuerst committed:
!!! BUGFIX: DocumentUriProjection doesn't respect fallback nodes; NodePropertiesWereSet contains affectedDimensionSpacePoints

Extending the NodePropertiesWereSet event is useful for some projections
which are not interested in OriginDimensionSpacePoints, but need to know
where a property value is applied (like the DocumentUriPathProjection).

This change is BREAKING because it updates the event payload of
NodePropertiesWereSet. To update your event store, run:

./flow migrateEvents:fillAffectedDimensionSpacePointsInNodePropertiesWereSet

We also fix the DocumentUriPathProjection alongside.

Resolves: #4265
Resolves: #4256

skurfuerst committed:
ensure Neos Behat tests run through again

dlubitz commented:
Looks good by reading. Will try it within next days. 
Just one test is failing due to the changes.

ahaeslich commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Event/NodePropertiesWereSet.php:
```diff
@@ -41,7 +42,9 @@ final class NodePropertiesWereSet implements
-        public readonly OriginDimensionSpacePoint $originDimensionSpacePoint, // TODO
+        public readonly OriginDimensionSpacePoint $originDimensionSpacePoint,
+        /** the covered dimension space points for this modification - i.e. where this change is visible */
+        public readonly DimensionSpacePointSet $affectedDimensionSpacePoints,
```
There are two other places in the current code base where `NodePropertiesWereSet` is initialized, which are not addressed in this MR:

- `Neos\ContentRepository\StructureAdjustment\Adjustment\PropertyAdjustment->publishNodePropertiesWereSet()`
- `Neos\ContentRepository\LegacyNodeMigration\NodeDataToEventsProcessor->createNodeVariant()`

skurfuerst replied:
good catch :) thanks!

skurfuerst replied:
done

ahaeslich commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/FeatureContext.php:
```diff
@@ -111,11 +120,80 @@ public function __construct()
         $this->objectManager = self::$bootstrap->getObjectManager();
         $this->environment = $this->objectManager->get(Environment::class);
 
+        $this->nodeAuthorizationService = $this->objectManager->get(AuthorizationService::class);
```
This is a duplicate of the next line
```suggestion
```

skurfuerst replied:
done

skurfuerst committed:
fix remaining places where NodePropertiesWereSet was wrongly used, fix tests

skurfuerst committed:
fix duplicate line

skurfuerst commented:
let's hope CI runs through now, then this would be merge ready IMHO. Tests locally run though :)

Thanks everybody!

skurfuerst committed:
!!! BUGFIX: DocumentUriPathProjection produces false uris on replay

If this case hit you, after this is merged, you can run:

   ./flow cr:replay documenturipath

to fix the problem.

Resolves: #4253

ahaeslich commented:
> let's hope CI runs through now

It won't due to the recent dependency changes in https://github.com/neos/flow-development-collection/commits/9.0

ahaeslich reviewed (approved):
Tested the changes in our project 👍 💖

skurfuerst committed:
Merge remote-tracking branch 'origin/9.0' into 90-documenturiprojection-respect-fallback-nodes
